File: homework4/problem3.c
#include <stddef.h>
#include <stdio.h>
#include <limits.h>

int array_minimum(int* minArray, size_t arrayLength, int minimumNumber) {
    int lowNumber = minimumNumber;
    if (arrayLength == 0) {
        return minimumNumber;
    } else if (minArray[arrayLength - 1] < minimumNumber){
        lowNumber = minArray[arrayLength - 1];
    }
    return array_minimum(minArray, arrayLength - 1, lowNumber);
}

int main() {
    int array[3] = {1, 2, 3};
    printf("%d\n", array_minimum(array, 3, INT_MAX)); // prints 1
}

File: homework4/problem3.go
package main

import "fmt"
import "math"

func main() {
    var array = []int{1, 2, 3}
    fmt.Println(arrayMinimum(array, math.MaxInt32))
}

func arrayMinimum(minArray []int, minimumNumber int) int {
    minArrayLength := len(minArray)
    smallestNumber := minimumNumber
    if minArrayLength == 0 {
        return minimumNumber
    } else if minArray[0] < smallestNumber {
        smallestNumber = minArray[0]
    }
    return arrayMinimum(minArray[1:], smallestNumber)
}
